### PR TITLE
Guidance feature for Collision Prevention

### PR DESCRIPTION
**Pull request by baumanta**
Add guidance feature to the collision prevention library. Enable the algorithm to slightly change the setpoint direction to help navigate in cluttered environments and fly through narrow gaps.

SITL and flight tested

### CODE REVIEW

**Comment by hamishwillee** [comment]
@baumanta Let me know if this needs any docs - when you're ready to submit.

**Comment by baumanta** [comment]
The correction for pitch was quite minimal, so I don't think that will impair the breaking maneuver. But doing the adaption had some negative effects on recognizing valid sensor data. E.g. if the sensor registers no obstacle, it sends a value just above the sensor range. Then we know, we should limit flight speed, such we can break at that distance, but we do not actually have any sensor data. If that sensor overlaps with a longer range sensor, we will want to use the data of that one, because the other value is not actually valid. If we now do the pitch adaption and the value gets smaller to the point that is is inside sensor range, we cannot know anymore whether it is an actual obstacle reading or not. We could solve that by using another array stating whether data is valid or not, but I don't think it is worth the effort

**Comment by mhkabir** [comment]
How about we perform the pitch correction in the distance sensor callback, and only if the sensor is within range?

When testing in simulation, I had found that the pitch correction helped greatly when accelerating (pitch down) towards obstacles.

**Comment by baumanta** [comment]
@mhkabir with `distance sensor callback` do you refer to the mavlink receiver? that would actually be an option that should work I think, the question is just if that would be ok for all other modules etc. using this data

**Comment by baumanta** [comment]
This has been tested extensively and flies quite nice. here a link to some of the flight logs:
https://logs.px4.io/plot_app?log=a4a2c795-3993-4ea5-bd0e-f9272ea7c9ee
https://logs.px4.io/plot_app?log=54a8fe23-5832-44b1-8337-e313cdf8eb11
https://logs.px4.io/plot_app?log=16cf619a-3855-4f0c-838d-e28c7c4f0f6e

File: src/lib/CollisionPrevention/CollisionPreventionTest.cpp [hunk 1]
```diff
@@ -404,16 +414,16 @@ TEST_F(CollisionPreventionTest, outsideFOV)
...
-			EXPECT_LT(modified_setpoint.norm(), 10.f);
+		//THEN: if the resulting setpoint demands velocities bigger zero, it must lie inside the FOV
+		float setpoint_length = modified_setpoint.norm();
 
-		} else {
-			// THEN: outside the FOV the setpoint should be clamped to zero
-			EXPECT_FLOAT_EQ(modified_setpoint.norm(), 0.f);
+		if (setpoint_length > 0.f) {
+			matrix::Vector2f setpoint_dir = modified_setpoint / setpoint_length;
+			float sp_angle_body_frame = atan2(setpoint_dir(1), setpoint_dir(0));
```
Comment (dagar) [inline src/lib/CollisionPrevention/CollisionPreventionTest.cpp]:
```suggestion
			float sp_angle_body_frame = atan2f(setpoint_dir(1), setpoint_dir(0));
```

File: src/lib/CollisionPrevention/CollisionPreventionTest.cpp [hunk 2]
```diff
@@ -804,3 +814,249 @@ TEST_F(CollisionPreventionTest, addObstacleSensorData_resolution_offset)
...
+
+	for (int i = 0; i < 7 ; i++) {
+		obstacle_msg.distances[i] = 500;
+	}
+
+	obstacle_msg.distances[2] = 1000;
+
+	//define setpoint
+	matrix::Vector2f setpoint_dir(1, 0);
+	float sp_angle_body_frame = matrix::atan2(setpoint_dir(1), setpoint_dir(0)) - vehicle_yaw_angle_rad;
```
Comment (dagar) [inline src/lib/CollisionPrevention/CollisionPreventionTest.cpp]:
```suggestion
	float sp_angle_body_frame = matrix::atan2f(setpoint_dir(1), setpoint_dir(0)) - vehicle_yaw_angle_rad;
```

File: src/lib/CollisionPrevention/CollisionPreventionTest.cpp [hunk 3]
```diff
@@ -804,3 +814,249 @@ TEST_F(CollisionPreventionTest, addObstacleSensorData_resolution_offset)
...
+		obstacle_msg.distances[i] = 500;
+	}
+
+	obstacle_msg.distances[1] = 1000;
+	obstacle_msg.distances[2] = 1000;
+	obstacle_msg.distances[3] = 1000;
+
+	//define setpoint
+	matrix::Vector2f setpoint_dir(1, 0);
+	float sp_angle_body_frame = matrix::atan2(setpoint_dir(1), setpoint_dir(0)) - vehicle_yaw_angle_rad;
```
Comment (dagar) [inline src/lib/CollisionPrevention/CollisionPreventionTest.cpp]:
```suggestion
	float sp_angle_body_frame = matrix::atan2f(setpoint_dir(1), setpoint_dir(0)) - vehicle_yaw_angle_rad;
```

File: src/lib/CollisionPrevention/CollisionPrevention.cpp [hunk 4]
```diff
@@ -223,14 +262,64 @@ void CollisionPrevention::_addDistanceSensorData(distance_sensor_s &distance_sen
...
+		mean_dist = mean_dist / (2.f * filter_size + 1.f);
+		float deviation_cost = col_prev_d * 50.f * std::abs(i - sp_index_original);
+		float bin_cost = deviation_cost - mean_dist - _obstacle_map_body_frame.distances[bin];
+
+		if (bin_cost < best_cost && _obstacle_map_body_frame.distances[bin] != UINT16_MAX) {
+			best_cost = bin_cost;
+
+			float angle = math::radians((float)bin * INTERNAL_MAP_INCREMENT_DEG + _obstacle_map_body_frame.angle_offset);
+			angle  = wrap_2pi(vehicle_yaw_angle_rad + angle);
+			setpoint_dir = {cos(angle), sin(angle)};
```
Comment (dagar) [inline src/lib/CollisionPrevention/CollisionPrevention.cpp]:
```suggestion
			setpoint_dir = {cosf(angle), sinf(angle)};
```

File: src/lib/CollisionPrevention/CollisionPreventionTest.cpp (hunk 2 above)
Comment (jkflying) [reply src/lib/CollisionPrevention/CollisionPreventionTest.cpp]: Things prefixed with `matrix::` already get converted to float due to the funky macros over here https://github.com/PX4/Matrix/blob/973999a4d3c6d4d85bf0153230974fd9571f7846/matrix/stdlib_imports.hpp#L33-L61 , but let's rather be explicit

**Comment by jkflying** [comment]
Closed in favor of #13017

**Pull request closed without merging**